File: tests/qa/legacy_python/test_endpoints.py
import urllib.request
import json
import time

BASE_URL = "http://localhost:7331"

def test_write_neuron():
    print("Testing POST /neuron...")
    payload = {
        "raw_payload": "Hello from API Tests!",
        "vector_data": [0.1] * 16,
        "model_creator_hash": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
        "payload_type": "text",
        "dna": None,
        "adjacency": None
    }
    
    req = urllib.request.Request(f"{BASE_URL}/neuron", method="POST")
    req.add_header("Content-Type", "application/json")
    
    try:
        data = json.dumps(payload).encode("utf-8")
        with urllib.request.urlopen(req, data=data) as response:
            res_data = json.loads(response.read().decode())
            print(f"SUCCESS: {res_data}")
            return res_data["neuron_id"]
    except Exception as e:
        print(f"FAILED: {e}")
        return None

def test_query_neuron(neuron_id):
    print("Testing POST /query (CNQL)...")
    payload = {
        "cnql": "find *"
    }
    
    req = urllib.request.Request(f"{BASE_URL}/query", method="POST")
    req.add_header("Content-Type", "application/json")
    
    try:
        data = json.dumps(payload).encode("utf-8")
        with urllib.request.urlopen(req, data=data) as response:
            res_data = json.loads(response.read().decode())
            found = any(n["neuron"]["id"] == neuron_id for n in res_data)
            if found:
                print(f"SUCCESS: Neuron {neuron_id} found in query results!")
            else:
                print(f"FAILED: Neuron {neuron_id} not found in query results.")
    except Exception as e:
        print(f"FAILED: {e}")

if __name__ == "__main__":
    print("Starting API Tests...")
    nid = test_write_neuron()
    if nid:
        time.sleep(1) # wait for WAL flush if any
        test_query_neuron(nid)

File: tests/integration/tests/test_endpoints.rs
use cluaizd_types::NeuronId;
use serde_json::json;

mod common;

#[tokio::test]
async fn test_write_and_query_neuron() {
    common::wait_for_server().await;
    let client = common::get_client();

    println!("Testing POST /neuron...");
    
    // Create dummy payload matching Python test
    let vector_data = vec![0.1; 16];
    let payload = json!({
        "raw_payload": "Hello from Native Rust API Tests!",
        "vector_data": vector_data,
        "model_creator_hash": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
        "payload_type": "text",
        "dna": null,
        "adjacency": null
    });

    let res = client.post(&format!("{}/neuron", common::BASE_URL))
        .json(&payload)
        .send()
        .await
        .expect("Failed to send POST request");

    assert!(res.status().is_success(), "Failed to write neuron, status: {}", res.status());

    let res_json: serde_json::Value = res.json().await.expect("Failed to parse response JSON");
    println!("SUCCESS: {}", res_json);

    let neuron_id_str = res_json["neuron_id"].as_str().expect("No neuron_id in response");
    
    // Wait briefly for WAL flush (Durability::Lite)
    tokio::time::sleep(std::time::Duration::from_millis(100)).await;

    println!("Testing POST /query (CDQL)...");
    
    let query_payload = json!({
        "cdql": "find *"
    });

    let res = client.post(&format!("{}/query", common::BASE_URL))
        .json(&query_payload)
        .send()
        .await
        .expect("Failed to send query request");

    assert!(res.status().is_success(), "Failed to query neurons, status: {}", res.status());

    let results: Vec<serde_json::Value> = res.json().await.expect("Failed to parse query results");
    
    let found = results.iter().any(|r| {
        r["neuron"]["id"].as_str().unwrap_or("") == neuron_id_str
    });

    assert!(found, "Neuron {} not found in query results!", neuron_id_str);
    println!("SUCCESS: Neuron {} found in query results!", neuron_id_str);
}
